{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "346a0040",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "69885461-25f7-44ec-a16d-0a085237f016",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "import re\n",
    "from pathlib import Path\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from matplotlib_venn import venn2\n",
    "from tqdm.auto import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "45077fbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Config\n",
    "\n",
    "# Main directories\n",
    "\n",
    "project_dir = Path(\"/path/to/carmen-analysis\")\n",
    "\n",
    "data_dir = project_dir.joinpath(\"data\")\n",
    "figures_dir = project_dir.joinpath(\"figures\")\n",
    "\n",
    "# data_pub_dir = data_dir.joinpath(\"to-be-published\")\n",
    "\n",
    "data_subs_dir = data_dir.joinpath(\"subsidiary-files\")\n",
    "\n",
    "figures_gen_dir = figures_dir.joinpath(\"script-generated\")\n",
    "figures_main_dir = figures_gen_dir.joinpath(\"main-panels\")\n",
    "\n",
    "random_pep_dir = data_subs_dir.joinpath(\"random-peptides\")\n",
    "\n",
    "# Core source files\n",
    "\n",
    "# Haplotype frequencies from NMDP Registry Haplotype Frequencies database\n",
    "freqs_dir = data_dir.joinpath(\"nmdp-hla-frequencies\")\n",
    "freqs_pop_desc_file = freqs_dir.joinpath(\"populations-description.csv\")\n",
    "freqs_pop_desc_col_pop_code = \"Population_code\"  # column name\n",
    "freqs_pop_desc_col_pop_group = \"Population_group\"  # column name\n",
    "freqs_pop_desc_col_desc = \"Description\"  # column name\n",
    "freqs_pop_desc_col_a_b_drb1 = \"A_B_DRB1_count\"  # column name\n",
    "freqs_abc_org_file = freqs_dir.joinpath(\"A~C~B.xlsx\")\n",
    "\n",
    "# Other\n",
    "\n",
    "A4_width = 8.27\n",
    "A4_height = 11.69\n",
    "\n",
    "plt.rcParams[\"pdf.fonttype\"] = 42\n",
    "plt.rcParams[\"ps.fonttype\"] = 42\n",
    "plt.rcParams[\"svg.fonttype\"] = \"none\"\n",
    "\n",
    "# Setting random seed\n",
    "random_seed = 42\n",
    "\n",
    "random.seed(random_seed)\n",
    "np.random.seed(random_seed)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2d1bb6e-dc95-455e-98b0-383644b3b566",
   "metadata": {},
   "source": [
    "# Load Populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "11044680-2d87-4abe-a980-932499e37d47",
   "metadata": {},
   "outputs": [],
   "source": [
    "NMDP_FOLDER = freqs_dir\n",
    "\n",
    "\n",
    "def load_populations():\n",
    "    filename = freqs_pop_desc_file\n",
    "    df = pd.read_csv(filename)\n",
    "    df = df.rename(\n",
    "        columns={\n",
    "            freqs_pop_desc_col_pop_code: \"Short\",\n",
    "            freqs_pop_desc_col_pop_group: \"Broad\",\n",
    "            freqs_pop_desc_col_desc: \"Full\",\n",
    "            freqs_pop_desc_col_a_b_drb1: \"Cnt\",\n",
    "        }\n",
    "    )\n",
    "    return df.set_index(\"Short\")\n",
    "\n",
    "\n",
    "def load_haplotype_freqs(load_mhc_2=False):\n",
    "    filename = (\n",
    "        os.path.join(NMDP_FOLDER, \"A~C~B~DRB3-4-5~DRB1~DQB1.xlsx\")\n",
    "        if load_mhc_2\n",
    "        else freqs_abc_org_file\n",
    "    )\n",
    "\n",
    "    df = pd.read_excel(filename)\n",
    "\n",
    "    if load_mhc_2:\n",
    "        df[\"haplotype\"] = df.apply(\n",
    "            lambda row: f\"{trim_hla_name(row.A)}-{trim_hla_name(row.B)}-{trim_hla_name(row.C)}\"\n",
    "            f\"-{trim_hla_name(row['DRB3-4-5'])}-{trim_hla_name(row['DRB1'])}-{trim_hla_name(row['DQB1'])}\",\n",
    "            axis=1,\n",
    "        )\n",
    "    else:\n",
    "        df[\"haplotype\"] = df.apply(\n",
    "            lambda row: f\"{trim_hla_name(row.A)}-{trim_hla_name(row.B)}-{trim_hla_name(row.C)}\",\n",
    "            axis=1,\n",
    "        )\n",
    "\n",
    "    df = df.set_index(\"haplotype\")\n",
    "    return df\n",
    "\n",
    "\n",
    "def trim_hla_name(mhc):\n",
    "    if mhc in (\"DRBX*NNNN\"):\n",
    "        result = \"None\"\n",
    "    else:\n",
    "        h = re.findall(r\"^(A|B|C|DRB1|DRB3|DRB4|DRB5|DQB1)\\*(\\d+):(\\d+)[gQNL]*$\", mhc)[\n",
    "            0\n",
    "        ]\n",
    "        if h[0] in (\"A\", \"B\", \"C\"):\n",
    "            result = (\n",
    "                f\"{h[0]}*{h[1]}:{h[2]}\".replace(\"g\", \"\")\n",
    "                .replace(\"Q\", \"\")\n",
    "                .replace(\"N\", \"\")\n",
    "                .replace(\"L\", \"\")\n",
    "            )\n",
    "        else:\n",
    "            result = f\"{h[0]}*{h[1]}:{h[2]}\".replace(\"g\", \"\")\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5a33e47c-b55a-412e-ae46-d9967068f316",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_populations = load_populations()\n",
    "df_haplotype_freqs = load_haplotype_freqs()\n",
    "df_populations_broad = df_populations[[\"Broad\", \"Cnt\"]].groupby(\"Broad\").sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d34cf66",
   "metadata": {},
   "source": [
    "# Generate Results For Haplotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a52b2771",
   "metadata": {},
   "source": [
    "## Select Most Common Haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1aacc16b",
   "metadata": {},
   "outputs": [],
   "source": [
    "groups = [\"AFA\", \"API\", \"CAU\", \"HIS\", \"NAM\"]\n",
    "coverage = 0.95\n",
    "\n",
    "dfs = {}\n",
    "\n",
    "alleles = set()\n",
    "\n",
    "for group in tqdm(groups):\n",
    "    df = df_haplotype_freqs.sort_values(f\"{group}_freq\", ascending=False)\n",
    "    df[\"cumulative\"] = df[f\"{group}_freq\"].cumsum()\n",
    "    dfs[group] = df[df.cumulative <= coverage]\n",
    "    additional_row = df[df.cumulative > coverage].iloc[0]\n",
    "    dfs[group] = pd.concat([dfs[group], additional_row.to_frame().T])\n",
    "\n",
    "    for idx, row in dfs[group].iterrows():\n",
    "        for mhc in [row.A, row.B, row.C]:\n",
    "            alleles.add(f\"HLA-{mhc.replace('g', '')}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de918378",
   "metadata": {},
   "source": [
    "## Run The Ranking Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c0d4520",
   "metadata": {},
   "source": [
    "Here, we utilize part of the [CAPE](https://github.com/hcgasser/CAPE) framework created by [Gasser et al.](https://www.immunoinformaticsjournal.com/article/S2667-1190(24)00005-3/fulltext), specifically the `tools/MHC-I_rank_peptides.py` script described in [this section](https://github.com/hcgasser/CAPE#prepare-the-mhc-class-1-position-weight-matrix-predictor). The script predicts antigenicity for every generated peptide and selected allele.\n",
    "\n",
    "First, generate command lines to be used with that framework by running the bellow cell and saving its output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cff2ae78-fc12-4d05-b957-3b24fb448bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "text = \"\"\n",
    "for _MHC in alleles:\n",
    "    text += f\"MHC-I_rank_peptides.py --peptides_per_length 1000000 --alleles {_MHC} --tasks rank --output {random_pep_dir}\\n\"\n",
    "\n",
    "print(text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f096dda",
   "metadata": {},
   "source": [
    "Now, create a separate programming environment for the [CAPE](https://github.com/hcgasser/CAPE) framework by following its [installation instructions](https://github.com/hcgasser/CAPE?tab=readme-ov-file#installation).\n",
    "\n",
    "After setting up the container, run the above command lines within it."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c905b6e",
   "metadata": {},
   "source": [
    "## Prepare Binding Peptides For Each Allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9972d7e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ignore = [\"HLA-A*15:150\"]\n",
    "\n",
    "path = random_pep_dir\n",
    "\n",
    "MHCs = {\"A\": [], \"B\": [], \"C\": []}\n",
    "for filename in os.listdir(path):\n",
    "    matches = re.findall(r\"HLA-(\\w)\\_(\\d+):(\\d+).csv\", filename)\n",
    "    if len(matches) > 0:\n",
    "        matches = matches[0]\n",
    "        mhc = f\"HLA-{matches[0]}*{matches[1]}:{matches[2]}\"\n",
    "        if mhc not in ignore:\n",
    "            MHCs[matches[0]].append(mhc)\n",
    "\n",
    "limit = 0.02\n",
    "\n",
    "sets = {}\n",
    "\n",
    "for mhc in tqdm(MHCs[\"A\"] + MHCs[\"B\"] + MHCs[\"C\"]):\n",
    "    if mhc not in sets:\n",
    "        peptides = set()\n",
    "        for nr, line in enumerate(\n",
    "            open(os.path.join(path, f'{mhc.replace(\"*\", \"_\")}.csv'))\n",
    "        ):\n",
    "            line = line.strip()\n",
    "            if nr > 0:\n",
    "                peptide, el_rank = line.split(\",\")\n",
    "                if float(el_rank) <= limit:\n",
    "                    peptides.add(peptide)\n",
    "\n",
    "        sets.update({mhc: peptides})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8537cd17-2f47-45a6-8d7f-7dc34faa4187",
   "metadata": {},
   "source": [
    "# Figure 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dde8c4bf",
   "metadata": {},
   "source": [
    "## Population Antigenicity Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0018dfa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_immuno_peptidome_overlaps(population_1, population_2, N, df_haplotype_freqs):\n",
    "    population = [population_1, population_2]\n",
    "\n",
    "    df = {\n",
    "        person: {\n",
    "            chromosome: df_haplotype_freqs.sample(\n",
    "                weights=f\"{population[person]}_freq\",\n",
    "                n=N,\n",
    "                replace=True,\n",
    "            )\n",
    "            for chromosome in range(2)\n",
    "        }\n",
    "        for person in range(2)\n",
    "    }\n",
    "\n",
    "    person_alleles = [[], []]\n",
    "    only_1 = []\n",
    "    only_2 = []\n",
    "    intersection = []\n",
    "\n",
    "    missing = 0\n",
    "    for pair in tqdm(range(N), leave=False, desc=str(population)):\n",
    "        for person in range(2):\n",
    "            _alleles = []\n",
    "            for chromosome in range(2):\n",
    "                match = re.findall(\n",
    "                    r\"A\\*(\\d+)\\:(\\d+)-B\\*(\\d+)\\:(\\d+)-C\\*(\\d+)\\:(\\d+)\",\n",
    "                    df[person][chromosome].iloc[pair].name,\n",
    "                )[0]\n",
    "                _alleles += [\n",
    "                    f\"HLA-A*{match[0]}:{match[1]}\",\n",
    "                    f\"HLA-B*{match[2]}:{match[3]}\",\n",
    "                    f\"HLA-C*{match[4]}:{match[5]}\",\n",
    "                ]\n",
    "            person_alleles[person].append(_alleles)\n",
    "\n",
    "        if all(mhc in sets for mhc in person_alleles[0][-1]) and all(\n",
    "            mhc in sets for mhc in person_alleles[1][-1]\n",
    "        ):\n",
    "            peptides_1 = set().union(*[sets[mhc] for mhc in person_alleles[0][-1]])\n",
    "            peptides_2 = set().union(*[sets[mhc] for mhc in person_alleles[1][-1]])\n",
    "\n",
    "            _only_1 = len(peptides_1 - peptides_2)\n",
    "            _only_2 = len(peptides_2 - peptides_1)\n",
    "            _int = len(peptides_1 & peptides_2)\n",
    "\n",
    "            _total = _only_1 + _only_2 + _int\n",
    "\n",
    "            only_1.append(_only_1 / _total)\n",
    "            only_2.append(_only_2 / _total)\n",
    "            intersection.append(_int / _total)\n",
    "        else:\n",
    "            missing += 1\n",
    "\n",
    "    return np.array(only_1), np.array(only_2), np.array(intersection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbecd7df-9264-47fe-855b-1ea27c567e5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "N = 10000\n",
    "\n",
    "df_results = pd.DataFrame(\n",
    "    index=df_populations_broad.index, columns=df_populations_broad.index\n",
    ")\n",
    "\n",
    "for i, population_1 in enumerate(df_results.index):\n",
    "    for j, population_2 in enumerate(df_results.columns):\n",
    "        if i <= j:\n",
    "            df_results.at[population_1, population_2] = get_immuno_peptidome_overlaps(\n",
    "                population_1, population_2, N, df_haplotype_freqs\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e402d2ce-00db-4867-905d-0b9e63521681",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_name = \"fig-5-population-antigenicity-comparison\"\n",
    "\n",
    "fig = plt.figure(figsize=(1.5 * A4_width, 1.5 * A4_width))\n",
    "\n",
    "n_grps = len(df_results.index)\n",
    "\n",
    "all_axes = []\n",
    "gs = mpl.gridspec.GridSpec(\n",
    "    n_grps + 1,\n",
    "    n_grps + 1,\n",
    "    height_ratios=[1] * (n_grps + 1),\n",
    "    width_ratios=[1] * (n_grps + 1),\n",
    "    wspace=0.0,\n",
    "    hspace=0.0,\n",
    ")\n",
    "\n",
    "for i, population_1 in enumerate(df_results.index):\n",
    "    for j, population_2 in enumerate(df_results.columns):\n",
    "        if i <= j:\n",
    "            r_1, c_1 = i, j + 1\n",
    "            r_2, c_2 = j + 1, i\n",
    "\n",
    "            ax = fig.add_subplot(gs[r_1, c_1])\n",
    "            only_1, only_2, intersection = df_results.at[population_1, population_2]\n",
    "\n",
    "            o1 = np.mean(only_1)\n",
    "            o2 = np.mean(only_2)\n",
    "            ints = np.mean(intersection)\n",
    "\n",
    "            venn = venn2((o1, o2, ints), set_labels=[population_1, population_2], ax=ax)\n",
    "            venn.get_label_by_id(\"10\").set_text(f\"{o1*100:.1f}%\")\n",
    "            venn.get_label_by_id(\"01\").set_text(f\"{o2*100:.1f}%\")\n",
    "            venn.get_label_by_id(\"11\").set_text(f\"{ints*100:.1f}%\")\n",
    "\n",
    "            ax = fig.add_subplot(gs[r_2, c_2])\n",
    "            sns.kdeplot(intersection, ax=ax, color=\"#AAAA00\")\n",
    "\n",
    "            ax.text(x=0.5, y=0.4, s=f\"{population_1} {population_2}\")\n",
    "\n",
    "            ax.spines[\"top\"].set_visible(False)\n",
    "            ax.spines[\"right\"].set_visible(False)\n",
    "            ax.spines[\"left\"].set_visible(False)\n",
    "            ax.get_yaxis().set_visible(False)\n",
    "            ax.set_xticks([0, 0.25, 0.5, 0.75, 1.0])\n",
    "            ax.set_xlim((0.0, 1.0))\n",
    "            ax.set_xticklabels([None, \"25%\", \"50%\", \"75%\", None])\n",
    "            ax.axvline(ints, linestyle=\"dashed\", color=\"#AAAA00\")\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "file_name = figures_main_dir.joinpath(f\"{fig_name}.pdf\")\n",
    "fig.savefig(file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4503269a-0384-49ef-89e6-7a0138748f08",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "carmen-analysis-2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
